presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java

```diff
@@ -610,19 +610,22 @@ public void close()
                 .collect(Collectors.toMap(Entry::getKey, entry -> utf8Slice(entry.getValue())));
 
         List<ColumnStatistics> unencryptedStats = new ArrayList<>();
-        Map<Integer, List<Slice>> encryptedStats = new HashMap<>();
+        Map<Integer, Map<Integer, Slice>> encryptedStats = new HashMap<>();
         addStatsRecursive(fileStats, 0, new HashMap<>(), unencryptedStats, encryptedStats);
         Optional<DwrfEncryption> dwrfEncryption;
         if (dwrfWriterEncryption.isPresent()) {
             ImmutableList.Builder<EncryptionGroup> encryptionGroupBuilder = ImmutableList.builder();
             List<WriterEncryptionGroup> writerEncryptionGroups = dwrfWriterEncryption.get().getWriterEncryptionGroups();
             for (int i = 0; i < writerEncryptionGroups.size(); i++) {
                 WriterEncryptionGroup group = writerEncryptionGroups.get(i);
+                int groupId = i;
                 encryptionGroupBuilder.add(
                         new EncryptionGroup(
                                 group.getNodes(),
                                 Optional.empty(), // reader will just use key metadata from the stripe
-                                encryptedStats.get(i)));
+                                group.getNodes().stream()
+                                        .map(nodeId -> encryptedStats.get(groupId).get(nodeId))
```

Comment:
You can optimize this piece a bit by extracting and moving `encryptedStats.get(groupId)` outside to line 621.